### PR TITLE
[SM100] Hdim 256 optimized

### PR DESCRIPTION
**Pull request by simon-veitner-redhat**
## Purpose

Re-enables FA4 for `head_size=256` on Blackwell, which was temporarily disabled in
https://github.com/vllm-project/vllm/pull/52050 because the SM100 2-CTA hd256 kernel rejected
`seqused_q/seqused_k` — something vLLM's decoder attention always supplies.

That gap is now closed upstream in https://github.com/vllm-project/flash-attention/pull/180,
which adds forward `seqused` support along with hd256 optimizations. This PR reverts #52050 and
integrates the kernel properly, with a fallback to FA2 for everything it cannot serve:

- The kernel reads paged KV through TMA only, with one page per 128-token tile, so the backend
  now advertises an exact KV cache block size of 128 for hd256. That 128 is chosen automatically
  when the user does not pin one; a user-pinned `--block-size` that is not a multiple of 128 makes
  FLASH_ATTN ineligible for hd256 (an error if the backend was requested explicitly).
- `forward()` page-aligns `max_seqlen_k` and slices the block table to match; the over-hang is
  masked out by `seqused_k`.
- The kernel has no SplitKV support, so `num_splits=1` is pinned for this path, and cascade
  attention is disabled.
- Selection falls back to FA2 for attention sinks, logits soft capping, mm_prefix, R-SWA and DCP.
  In practice Gemma-2 (soft capping) and multimodal Gemma-3 (mm_prefix) keep using FA2, while
  Gemma-1, text-only Gemma-3 and PaliGemma/ColPali get FA4.
- The hd256 kernel is opt-in (`supports_fa4_hd256`), so callers that do not build the required
  call shape — ViT/`mm_encoder_attention`, turboquant, MLA prefill — keep resolving to FA2.
- Only `(head_size=256, head_size_v=256)` uses this kernel; `(head_size=256, head_size_v=128)`
  still takes generic FA4, exactly as it did before #52050.

Sliding window on the decoder path is enabled again, since the kernel handles it; encoder inputs
(dense `cu_seqlens_k`, no `seqused_k`) still fall back to FA2.

## Test Plan

```bash
pytest tests/kernels/attention/test_attention_selector.py
pytest tests/kernels/attention/test_flash_attn.py -k fa4_hd256
pytest tests/kernels/attention/test_flash_attn.py

# reproducer from #52050
pytest tests/models/multimodal/pooling/test_colpali.py::test_colpali_multimodal_text_query_image_docs
```

## Test Result

On B200 (SM100):

```
tests/kernels/attention/test_attention_selector.py            52 passed, 10 skipped
tests/kernels/attention/test_flash_attn.py -k fa4_hd256        8 passed
```

---

AI assistance was used for the vLLM-side changes. I reviewed every changed line and ran the
tests reported above.

### CODE REVIEW

File: vllm/v1/attention/backends/fa_utils.py [hunk 1]
```diff
@@ -209,6 +231,66 @@ def get_flash_attn_version(
...
+    """
+    if head_size != 256:
+        return False
+    if head_size_v is not None and head_size_v != 256:
+        return False
+    capability = current_platform.get_device_capability()
+    return capability is not None and capability.major in (10, 11)
+
+
+def _fa4_hd256_fallback_reason(
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/fa_utils.py:236]: We also need to fall back when `is_quantized_kv_cache(cache_config.cache_dtype)`

File: vllm/v1/attention/backends/flash_attn.py [hunk 2]
```diff
@@ -87,13 +89,36 @@ class FlashAttentionBackend(AttentionBackend):
...
+    def _get_fa4_hd256_block_size() -> int | None:
+        vllm_config = get_current_vllm_config_or_none()
+        if vllm_config is None or vllm_config.model_config is None:
+            return None
+
+        head_size = vllm_config.model_config.get_head_size()
+        # No kv_cache_block_size: this call is what *decides* it.
+        if (
+            uses_fa4_hd256_kernel(head_size)
+            and get_flash_attn_version(head_size=head_size, supports_fa4_hd256=True)
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py]: We also need to pass `head_size_v` (and set it to `None` in this parent class) so that DiffKV (child) works correctly

File: vllm/v1/attention/backends/flash_attn.py [hunk 3]
```diff
@@ -780,6 +829,11 @@ def update_draft_decode_metadata(self, metadata: FlashAttentionMetadata) -> None
         metadata.scheduler_metadata = self._store_scheduler_metadata(scheduler_metadata)
 
     def use_cascade_attention(self, *args, **kwargs) -> bool:
+        if self.fa4_hd256:
+            # The cascade path calls FA with the unsliced block table and a
+            # prefix length that is not page aligned, which the hd256 kernel
+            # rejects. Cascade is an optimization, so just turn it off.
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py]: Nit: clean up comment

File: vllm/v1/attention/backends/flash_attn.py [hunk 4]
```diff
@@ -823,11 +877,36 @@ def __init__(
...
+        # the page size is enforced at spec/group level (select_common_block_size).
         self.vllm_flash_attn_version = get_flash_attn_version(
             requires_alibi=alibi_slopes is not None,
             head_size=head_size,
             has_sinks=sinks is not None,
+            requires_softcap=bool(self.logits_soft_cap),
+            supports_fa4_hd256=True,
+        )
+        # Requests routed to FA4's dedicated Blackwell head_dim=256 kernel need
+        # the call-shape adjustments made in forward() below.
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py]: nit: clean up

File: vllm/v1/attention/backends/flash_attn.py [hunk 5]
```diff
@@ -823,11 +877,36 @@ def __init__(
...
+        self.fa4_hd256 = self.vllm_flash_attn_version == 4 and uses_fa4_hd256_kernel(
+            head_size
         )
+        if self.fa4_hd256 and not uses_kv_cache and sliding_window is not None:
+            # That kernel implements local attention only alongside
+            # seqused_q/seqused_k. The KV cache path always passes seqused_k,
+            # but encoder attention runs on dense cu_seqlens_k (e.g.
+            # google/embeddinggemma-300m: head_size=256, encoder-only, windowed).
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py]: nit: clean up

File: vllm/v1/attention/backends/flash_attn.py [hunk 6]
```diff
@@ -1083,6 +1163,16 @@ def forward(
                     )
                     causal = not has_window
 
+                num_splits = attn_metadata.max_num_splits
+                if self.fa4_hd256:
+                    # The kernel requires max_seqlen_k % page_size == 0, a block
+                    # table of exactly max_seqlen_k // page_size columns (the
+                    # over-hang is masked out by seqused_k), and no SplitKV.
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py]: nit: clean up

File: vllm/v1/attention/backends/flash_attn.py [hunk 7]
```diff
@@ -1416,7 +1506,8 @@ def _forward_encoder_attention(
             else None,
             k_descale=layer._k_scale.expand(descale_shape),  # type: ignore[operator]
             v_descale=layer._v_scale.expand(descale_shape),  # type: ignore[operator]
-            num_splits=1 if self.batch_invariant_enabled else 0,
+            # hd256: no SplitKV, so never let FA pick num_splits > 1.
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py]: nit: clean up

File: vllm/v1/attention/backends/flash_attn_diffkv.py [hunk 8]
```diff
@@ -87,6 +87,10 @@ def __init__(self, *args, **kwargs) -> None:
             head_size_v=FlashAttentionDiffKVBackend.head_size_v,
             has_sinks=self.sinks is not None,
         )
+        # The dedicated hd256 kernel needs head_size_v == head_size == 256 and
+        # the call-shape fixups in FlashAttentionImpl.forward, which this
+        # subclass's KV path does not apply.
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn_diffkv.py]: nit: clean up

File: vllm/v1/attention/backends/flash_attn.py [hunk 9]
```diff
@@ -409,6 +447,17 @@ def __init__(
         self.max_num_splits = 0  # No upper bound on the number of splits.
         self.aot_schedule = get_flash_attn_version() == 3
 
+        # FA4's dedicated Blackwell head_dim=256 kernel. self.block_size is
+        # this KV cache group's *actual* kernel page size.
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/flash_attn.py]: nit: clean up

File: vllm/v1/attention/backends/fa_utils.py [hunk 10]
```diff
@@ -209,6 +232,69 @@ def get_flash_attn_version(
...
+    head_size: int | None, head_size_v: int | None = None
+) -> bool:
+    """Whether FA4 would dispatch this layer to its dedicated hd256 kernel.
+
+    Mirrors ``use_dedicated_hd256_kernel`` in flash_attn/cute/interface.py:
+    ``arch // 10 in [10, 11] and head_dim == head_dim_v == 256``. head_size=256
+    with a different V head dim is not a shape FA4 supports on SM100 at all
+    (``_validate_head_dims`` asserts on it); the TMEM block above resolves it
+    to FA2.
+    """
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/fa_utils.py]: nit: single line docstring should be fine

File: vllm/v1/attention/backends/fa_utils.py [hunk 11]
```diff
@@ -6,9 +6,15 @@
 import vllm.envs as envs
 from vllm.logger import init_logger
 from vllm.platforms import current_platform
+from vllm.utils.torch_utils import is_quantized_kv_cache
 
 logger = init_logger(__name__)
 
+# FA4 serves (head_dim, head_dim_v) == (256, 256) on Blackwell with a dedicated
+# 2-CTA kernel (flash_attn/cute/sm100_hd256_2cta_fmha_forward.py) that reads
```
Comment (MatthewBonanni) [inline vllm/v1/attention/backends/fa_utils.py]: nit: not necessarily 2-CTA anymore, yeah?

File: tests/kernels/attention/test_attention_selector.py [hunk 12]
```diff
@@ -555,3 +558,202 @@ def test_flash_attn_accepts_handled_fp8_variants(
     # import order across earlier tests that patch vllm.platforms.current_platform.
     monkeypatch.setattr(fa_utils_mod.current_platform, "is_xpu", lambda: True)
     assert FlashAttentionBackend.supports_kv_cache_dtype(kv_cache_dtype)
+
+
+# FA4 dispatches head_size=256 on Blackwell to a dedicated 2-CTA kernel whose
```
Comment (MatthewBonanni) [inline tests/kernels/attention/test_attention_selector.py]: nit: not necessarily 2-CTA anymore, yeah?

**Review by MatthewBonanni — APPROVED** [review]
LGTM, thanks! Can you just clean up the comments that are self-evident? I marked some of the ones that I think can be removed.
